packages/yarnpkg-fslib/sources/ZipOpenFS.ts

```diff
@@ -721,24 +728,39 @@ export class ZipOpenFS extends BasePortableFakeFS {
     }
   }
 
-  private limitOpenFiles(max: number) {
+  private limitOpenFilesTimeout: NodeJS.Timeout | null = null;
+  private limitOpenFiles(max: number | null) {
     if (this.zipInstances === null)
       return;
 
-    let closeCount = this.zipInstances.size - max;
+    const now = Date.now();
+    let closeCount = max === null ? 0 : this.zipInstances.size - max;
 
-    for (const [path, zipFs] of this.zipInstances.entries()) {
-      if (closeCount <= 0)
-        break;
-
-      if (zipFs.hasOpenFileHandles())
+    for (const [path, {zipFs, time}] of this.zipInstances.entries()) {
+      if (zipFs.hasOpenFileHandles()) {
+        continue;
+      } else if (now - time >= this.maxAge) {
+        zipFs.saveAndClose();
+        this.zipInstances.delete(path);
+        closeCount -= 1;
+        continue;
+      } else if (max === null) {
         continue;
+      } else if (closeCount <= 0) {
+        break;
+      }
 
       zipFs.saveAndClose();
       this.zipInstances.delete(path);
-
       closeCount -= 1;
     }
+
+    if (this.limitOpenFilesTimeout === null && ((max === null && this.zipInstances.size > 0) || max !== null)) {
+      this.limitOpenFilesTimeout = setTimeout(() => {
+        this.limitOpenFilesTimeout = null;
+        this.limitOpenFiles(null);
+      }, 1000).unref();
```

Comment:
Would be better to use the traversal to find the minimal amount of time until the next garbage collection, and use that.